Gems/EMotionFX/Code/Source/Editor/Plugins/Ragdoll/JointSwingLimitManipulators.cpp

```diff
@@ -0,0 +1,245 @@
+    static constexpr float ManipulatorInverseScale = 1.0f / ManipulatorScale;
+
+    JointSwingLimitManipulators::JointSwingLimitManipulators()
+    {
+        m_adjustJointLimitCallback = AZStd::make_unique<PhysicsSetupManipulatorCommandCallback>(this, false);
+        EMStudio::GetCommandManager()->RegisterCommandCallback("AdjustJointLimit", m_adjustJointLimitCallback.get());
+    }
+
+    JointSwingLimitManipulators::~JointSwingLimitManipulators()
+    {
+        EMStudio::GetCommandManager()->RemoveCommandCallback(m_adjustJointLimitCallback.get(), false);
+    }
+
+    void JointSwingLimitManipulators::Setup(PhysicsSetupManipulatorData& physicsSetupManipulatorData)
+    {
+        m_physicsSetupManipulatorData = physicsSetupManipulatorData;
+
+        if (!m_physicsSetupManipulatorData.HasJointLimit())
+        {
+            return;
+        }
+
+        AzFramework::DebugDisplayRequestBus::BusPtr debugDisplayBus;
+        AzFramework::DebugDisplayRequestBus::Bind(debugDisplayBus, m_viewportId);
+        m_debugDisplay = AzFramework::DebugDisplayRequestBus::FindFirstHandler(debugDisplayBus);
+
+        // swing limit Y manipulator
+        m_swingYManipulator = AzToolsFramework::LinearManipulator::MakeShared(AZ::Transform::CreateIdentity());
+        m_swingYManipulator->SetAxis(AZ::Vector3::CreateAxisZ());
+        m_swingYManipulator->Register(EMStudio::g_animManipulatorManagerId);
+        {
+            AzToolsFramework::ManipulatorViews views;
+            views.emplace_back(AzToolsFramework::CreateManipulatorViewQuadBillboard(
+                AzFramework::ViewportColors::DefaultManipulatorHandleColor, AzFramework::ViewportConstants::DefaultManipulatorHandleSize));
+            m_swingYManipulator->SetViews(AZStd::move(views));
+        }
+
+        m_swingYManipulator->InstallLeftMouseDownCallback(
+            [this]([[maybe_unused]] const AzToolsFramework::LinearManipulator::Action& action)
+            {
+                BeginEditing();
+            });
+        m_swingYManipulator->InstallMouseMoveCallback(
+            [this](const AzToolsFramework::LinearManipulator::Action& action)
+            {
+                if (m_jointSwingLimitState.m_swingLimitY.has_value())
+                {
+                    float newSwingLimitY = ManipulatorScale * action.LocalPosition().GetZ();
+                    m_physicsSetupManipulatorData.m_jointConfiguration->SetPropertyValue(AZ::Name("SwingLimitY"), newSwingLimitY);
+                    AZStd::optional<float> validatedSwingLimitY =
+                        m_physicsSetupManipulatorData.m_jointConfiguration->GetPropertyValue(AZ::Name("SwingLimitY"));
+                    if (validatedSwingLimitY.has_value())
+                    {
+                        m_swingYManipulator->SetLocalPosition(
+                            AZ::Vector3(ManipulatorOffsetX, 0.0f, ManipulatorInverseScale * validatedSwingLimitY.value()));
+                    }
```

Comment:
question: I guess this is doing a set and a get because the setter it's validation and clamping the limit value? So the getter could returned a different value, right?

Reply:
yeah, that's right

Reply:
OK, no problem, it's a good idea to add a comment.